File: app/src/main/java/com/example/udemy_thecompleteandroid14developercourse_build100apps/MasteringAdapterAndGridView/ShapeAdapter.java
package com.example.udemy_thecompleteandroid14developercourse_build100apps.MasteringAdapterAndGridView;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.udemy_thecompleteandroid14developercourse_build100apps.R;

import java.util.ArrayList;

public class ShapeAdapter extends ArrayAdapter<Shape> {
    private ArrayList<Shape> shapesArrayList;
    Context context;

    public ShapeAdapter(ArrayList<Shape> shapesArrayList, Context context){
        super(context, R.layout.grid_item_layout, shapesArrayList);
        this.shapesArrayList = shapesArrayList;
        this.context = context;
    }
    //View Holder: Used to cache references to the views within an item layout
    public static class ViewHolder{
        TextView shapeName;
        ImageView shapeImage;
    }

    // GetView(): Used to create and return a view for a specific item in grid.


    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        // 1- Get the shape object for the current position
        Shape shapes = getItem(position);

        //2-Inflating Layout
        ViewHolder viewHolder;

        if (convertView == null){
            viewHolder = new ViewHolder();
            LayoutInflater inflater = LayoutInflater.from(getContext());

            convertView = inflater.inflate(
                    R.layout.grid_item_layout,
                    parent,
                    false
            );

            // Finding the views
            viewHolder.shapeName = (TextView) convertView.findViewById(R.id.shapeName);
            viewHolder.shapeImage = (ImageView) convertView.findViewById(R.id.shapeImage);

            convertView.setTag(viewHolder);

        } else {
            //A view went off-screen --> reuse it
            viewHolder = (ViewHolder) convertView.getTag();
        }

        viewHolder.shapeName.setText(shapes.getShapeName());
        viewHolder.shapeImage.setImageResource(shapes.getShapeImage());

        return convertView;
    }
}

File: app/src/main/java/com/example/udemy_thecompleteandroid14developercourse_build100apps/KotlinVolumeCalculatorApp/adapters/ShapeAdapter.kt
package com.example.udemy_thecompleteandroid14developercourse_build100apps.KotlinVolumeCalculatorApp.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import com.example.udemy_thecompleteandroid14developercourse_build100apps.KotlinVolumeCalculatorApp.models.Shape
import com.example.udemy_thecompleteandroid14developercourse_build100apps.R

class ShapeAdapter(context:Context, private val shapeList:List<Shape>)
    :ArrayAdapter<Shape>(context, 0, shapeList) {
        
        private class ShapeViewHolder{
            lateinit var shapeImageView:ImageView
            lateinit var shapeName:TextView
            
        }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var itemView = convertView
        val holder:ShapeViewHolder
        if (convertView == null){
            itemView = LayoutInflater.from(context).inflate(R.layout.grid_item_layout, parent, false)
            
            holder = ShapeViewHolder()
            holder.shapeImageView = itemView.findViewById(R.id.shapeImage)
            holder.shapeName = itemView.findViewById(R.id.shapeName)
            
            itemView.tag = holder
        }else {
            holder = itemView?.tag as ShapeViewHolder
        }
        
        val currentItem = shapeList[position]
        holder.shapeImageView.setImageResource(currentItem.shapeImage)
        holder.shapeName.text = currentItem.shapeName
        return itemView!!
    }
}